# Aplikasi NLP

In [ ]:
! pip install Sastrawi
from sklearn.pipeline import Pipeline
from Sastrawi.Stemmer.StemmerFactory import StemmerFactory     


[notice] A new release of pip is available: 24.0 -> 25.1.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [ ]:
pip install nltk#library untuk NLP

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.0 -> 25.1.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

import re # Regular Expression untuk membersihkan teks

import nltk
nltk.download('stopwords') # mengunduh kata kata umum seperti the,is,etc
nltk.download('punkt') # mengunduh tokenizer untuk memecah teks menjadi kata-kata
from nltk.corpus import stopwords # Mengimpor kumpulan stopwords untuk digunakan nanti saat membersihkan teks.
from nltk.stem import WordNetLemmatizer # mengimpor kata untuk lemmatisasi yaitu mengubah kata ke bentuk dasarnya

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\rifqi\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\rifqi\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!


In [ ]:
train_df = pd.read_csv(r'C:\Users\rifqi\Documents\Belajar Data Science\Streamlit\Data.csv', on_bad_lines='skip',sep=';')
train_df.head()# melihat 5 baris pertama

Unnamed: 0                                              title  \
0           0  Live from New York, it's a Trump-Clinton remat...   
1           1  Catalan separatists to lose majority in tight ...   
2           2  North Carolina governor concedes election to D...   
3           3  Draft Senate Iran legislation sets tough new U...   
4           4  California governor taps U.S. Representative B...   

                                                text  label  
0  NEW YORK (Reuters) - Veteran actor and frequen...      1  
1  BARCELONA (Reuters) - Catalonia s independence...      1  
2  WINSTON-SALEM, N.C. (Reuters) - North Carolina...      1  
3  WASHINGTON (Reuters) - Draft legislation respo...      1  
4  SACRAMENTO, Calif. (Reuters) - California Gove...      1

In [ ]:
train_df = train_df[['title', 'text', 'label']]#memilih hanya 3 kolom penting
train_df['content'] = train_df['title'] + " " + train_df['text']#membuat kolom baru bernama content dan diisi dengan kolom title dan text

train_df.head()

title  \
0  Live from New York, it's a Trump-Clinton remat...   
1  Catalan separatists to lose majority in tight ...   
2  North Carolina governor concedes election to D...   
3  Draft Senate Iran legislation sets tough new U...   
4  California governor taps U.S. Representative B...   

                                                text  label  \
0  NEW YORK (Reuters) - Veteran actor and frequen...      1   
1  BARCELONA (Reuters) - Catalonia s independence...      1   
2  WINSTON-SALEM, N.C. (Reuters) - North Carolina...      1   
3  WASHINGTON (Reuters) - Draft legislation respo...      1   
4  SACRAMENTO, Calif. (Reuters) - California Gove...      1   

                                             content  
0  Live from New York, it's a Trump-Clinton remat...  
1  Catalan separatists to lose majority in tight ...  
2  North Carolina governor concedes election to D...  
3  Draft Senate Iran legislation sets tough new U...  
4  California governor taps U.S. Representative B...

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer #mengubah teks menjadi angka berdasarkan tf-idf

X = train_df['content']# fitur input
y = train_df['label']# target atau label

tfidf_vectorizer = TfidfVectorizer(max_features=5000, ngram_range=(1, 2), stop_words='english')
# Hanya mengambil 5000 fitur lalu gunakan unigram(kata tunggal) dan bigram(pasangan kata) selanjutnya hapus stopwords dalam bahasa inggris
review_tf = tfidf_vectorizer.fit_transform(X.astype('U'))
#memastikasn semua data dalam tipe string unicode dan akan kita train dari teks untuk membuat tf-idf lalu mengubah  semua teks menjadi matriks angka berbentuk sparse matrix
review_tf

<Compressed Sparse Row sparse matrix of dtype 'float64'
	with 1071029 stored elements and shape (8117, 5000)>

In [ ]:
from sklearn.model_selection import train_test_split # untuk membagi data

X_train, X_test, y_train, y_test = train_test_split(review_tf, y, train_size = 0.8, test_size = 0.2, random_state = 42)
# review_tf	            Fitur input (hasil transformasi TF-IDF dari teks).
# y	                    Target label (seperti 0/1 untuk klasifikasi).
# train_size=0.8	    80% data digunakan untuk melatih model.
# test_size=0.2	20%     data digunakan untuk mengukur performa model.
# random_state=42	    Seed acak untuk memastikan hasil pembagian konsisten setiap kali kode dijalankan.

## pelatihan model klasifikasi dengan Random Forest

In [ ]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix

clf_rf = RandomForestClassifier().fit(X_train, y_train)# membuat model random forest
predicted_rf = clf_rf.predict(X_test)# model memprediksi label dari x_test dan hasilnya akan disimpan di predicted_rf

rf_accuracy = accuracy_score(y_test, predicted_rf)

print("Random Forest Accuracy: ", accuracy_score(y_test, predicted_rf)) # mengukur  seberapa banyak prediksi yang benar dibandingkan total data.
print("Random Forest Precision: ", precision_score(y_test, predicted_rf, average = "macro"))#  dari semua prediksi positif, berapa banyak yang benar.
print("Random Forest Recall: ", recall_score(y_test, predicted_rf, average = "macro"))#  dari semua data yang seharusnya positif, berapa banyak yang berhasil diprediksi benar
print("Random Forest F1 Score: ", f1_score(y_test, predicted_rf, average = "macro"))# komposisi dari presisi dan recall.

print(f'confusion matrix:\n {confusion_matrix(y_test, predicted_rf)}')# menampilkan tabel 2D, dimana Baris berisi label sebenarnya dan kolom berisi label yang diprediksi
print("----------------------------------------------------------")
print(classification_report(y_test, predicted_rf, zero_division = 0)) # menampilkan semua matriks  (precision, recall, F1, support)

Random Forest Accuracy:  0.979679802955665
Random Forest Precision:  0.9791260544006997
Random Forest Recall:  0.9806826749525921
Random Forest F1 Score:  0.9796316038515698
confusion matrix:
 [[756   3]
 [ 30 835]]
----------------------------------------------------------
              precision    recall  f1-score   support

           0       0.96      1.00      0.98       759
           1       1.00      0.97      0.98       865

    accuracy                           0.98      1624
   macro avg       0.98      0.98      0.98      1624
weighted avg       0.98      0.98      0.98      1624



### Penjelasan mengenai hasil evaluasi

#### Accuracy 

```
Random Forest Accuracy: 0.9796
```
- Dari 97.96% diprediksi dengan benar oleh model.

#### Precision

```
Random Forest Precision (macro): 0.9791
```

- Precision mengukur: dari semua prediksi positif, berapa persen yang benar-benar positif 
- Precision kelas 0: 0.96 --> 96% prediksi untuk kelas 0 benar
- Precision kelas 1: 1.00 --> 100% prediksi untuk kelas 1 benar

#### Recall

```
Random Forest Recall (macro): 0.98076
```
- Recall kelas 0: 1.00 --> Semua data kelas 0 berhasil diprediksi dengan benar
- Recall kelas 1: 0.97 --> 97% data kelas 1 berhasil diprediksi dengan benar

#### F1 Score

```
Random Forest F1 Score (macro): 0.9796
```
- F1 kelas 0: 0.98
- F1 kelas 1: 0.98

#### Confusion Matrix

```
[[756   3]
 [ 30 835]]
```

|               | Prediksi: 0 | Prediksi: 1 |
| ------------- | ----------- | ----------- |
| **Aktual: 0** | 756 (TP)    | 3 (FP)      |
| **Aktual: 1** | 30 (FN)     | 835 (TP)    |


#### Classification Report

- Support: jumlah data sebenarnya di setiap kelas
- Weighted avg: rata-rata dihitung berdasarkan banyaknya data tiap kelas
- Semua metrik sangat tinggi (≈0.98), menandakan model bekerja sangat baik di kedua kelas.





##  Validasi model Random Forest terhadap data baru

In [ ]:
# melakukan validasi model Random Forest terhadap data baru

valid_df = pd.read_csv(r'C:\Users\rifqi\Documents\Belajar Data Science\Streamlit\Data.csv', on_bad_lines='skip',sep=';')

valid_df = valid_df[['title', 'text', 'label']]
valid_df['content'] = valid_df['title'] + " " + valid_df['text']

validation_X = valid_df['content']
validation_tf = tfidf_vectorizer.transform(validation_X.astype('U'))

validation_predictions = clf_rf.predict(validation_tf)

if 'label' in valid_df.columns:
    validation_y = valid_df['label']
    print("Validation Accuracy: ", accuracy_score(validation_y, validation_predictions))
    print("Validation Precision: ", precision_score(validation_y, validation_predictions, average='macro'))
    print("Validation Recall: ", recall_score(validation_y, validation_predictions, average='macro'))
    print("Validation F1 Score: ", f1_score(validation_y, validation_predictions, average='macro'))
    print(f'Confusion matrix:/n{confusion_matrix(validation_y, validation_predictions)}')
    print(classification_report(validation_y, validation_predictions, zero_division=0))

valid_df['Predicted Label'] = validation_predictions
valid_df.head()

Validation Accuracy:  0.9959344585437969
Validation Precision:  0.9956858848373721
Validation Recall:  0.9961631061786684
Validation F1 Score:  0.995913303839843
Confusion matrix:/n[[3750    3]
 [  30 4334]]
              precision    recall  f1-score   support

           0       0.99      1.00      1.00      3753
           1       1.00      0.99      1.00      4364

    accuracy                           1.00      8117
   macro avg       1.00      1.00      1.00      8117
weighted avg       1.00      1.00      1.00      8117



title  \
0  Live from New York, it's a Trump-Clinton remat...   
1  Catalan separatists to lose majority in tight ...   
2  North Carolina governor concedes election to D...   
3  Draft Senate Iran legislation sets tough new U...   
4  California governor taps U.S. Representative B...   

                                                text  label  \
0  NEW YORK (Reuters) - Veteran actor and frequen...      1   
1  BARCELONA (Reuters) - Catalonia s independence...      1   
2  WINSTON-SALEM, N.C. (Reuters) - North Carolina...      1   
3  WASHINGTON (Reuters) - Draft legislation respo...      1   
4  SACRAMENTO, Calif. (Reuters) - California Gove...      1   

                                             content  Predicted Label  
0  Live from New York, it's a Trump-Clinton remat...                1  
1  Catalan separatists to lose majority in tight ...                1  
2  North Carolina governor concedes election to D...                1  
3  Draft Senate Iran legislation sets tough new U...                1  
4  California governor taps U.S. Representative B...                1

## Menyimpan model machine learning dan TF-IDF vectorizer ke dalam file

In [ ]:
# menyimpan model machine learning dan TF-IDF vectorizer ke dalam file, agar nanti bisa digunakan kembali tanpa harus dilatih ulang.

import pickle #  modul Python untuk menyimpan (serialize) dan memuat (deserialize) objek Python dalam bentuk file biner (.sav, .pkl, dll).

# Simpan TF-IDF Vectorizer
with open('tfidf_vectorizer.sav', 'wb') as f:# buka file untuk ditulis dalam mode binary (.sav)
    pickle.dump(tfidf_vectorizer, f)# simpan objek tfidf_vectorizer ke file tersebut
print("TF-IDF Vectorizer saved to 'tfidf_vectorizer.sav'")

# Simpan Random Forest Model
with open('random_forest_model.sav', 'wb') as f:
    pickle.dump(clf_rf, f)
print("Random Forest model saved to 'random_forest_model.sav'")


TF-IDF Vectorizer saved to 'tfidf_vectorizer.sav'
Random Forest model saved to 'random_forest_model.sav'


##  load model machine learning dan TF-IDF vectorizer dari file .sav

In [15]:
# Load TF-IDF Vectorizer
with open('tfidf_vectorizer.sav', 'rb') as f:
    tfidf_vectorizer_loaded = pickle.load(f)
print("TF-IDF Vectorizer loaded successfully.")

# Load Random Forest Model
with open('random_forest_model.sav', 'rb') as f:
    clf_rf_loaded = pickle.load(f)
print("Random Forest model loaded successfully.")

TF-IDF Vectorizer loaded successfully.
Random Forest model loaded successfully.


## Pengujian model yang sudah disimpan dan dimuat ulang

In [ ]:
# pengujian model yang sudah disimpan dan dimuat ulang (TF-IDF vectorizer dan Random Forest classifier), lalu digunakan untuk memprediksi data teks

test_df = pd.read_csv(r'C:\Users\rifqi\Documents\Belajar Data Science\Streamlit\Data.csv', on_bad_lines='skip',sep=';')
test_df = test_df[['title', 'text', 'label']]
test_df['content'] = test_df['title'] + " " + test_df['text']

# ransformasi Teks ke TF-IDF
test_X = test_df['content']
test_tf = tfidf_vectorizer_loaded.transform(test_X.astype('U'))

# Prediksi label
test_predictions = clf_rf_loaded.predict(test_tf)

# Evaluasi prediksi dengan metrik Accuracy, Precision, Recall, dan F1 Score
if 'label' in test_df.columns:
    test_y = test_df['label']
    print("Test Accuracy: ", accuracy_score(test_y, test_predictions))
    print("Test Precision: ", precision_score(test_y, test_predictions, average='macro'))
    print("Test Recall: ", recall_score(test_y, test_predictions, average='macro'))
    print("Test F1 Score: ", f1_score(test_y, test_predictions, average='macro'))
    print(f'Confusion matrix:/n{confusion_matrix(test_y, test_predictions)}')
    print(classification_report(test_y, test_predictions, zero_division=0))

test_df['Predicted Label'] = test_predictions
test_df.head()

Test Accuracy:  0.9959344585437969
Test Precision:  0.9956858848373721
Test Recall:  0.9961631061786684
Test F1 Score:  0.995913303839843
Confusion matrix:/n[[3750    3]
 [  30 4334]]
              precision    recall  f1-score   support

           0       0.99      1.00      1.00      3753
           1       1.00      0.99      1.00      4364

    accuracy                           1.00      8117
   macro avg       1.00      1.00      1.00      8117
weighted avg       1.00      1.00      1.00      8117



title  \
0  Live from New York, it's a Trump-Clinton remat...   
1  Catalan separatists to lose majority in tight ...   
2  North Carolina governor concedes election to D...   
3  Draft Senate Iran legislation sets tough new U...   
4  California governor taps U.S. Representative B...   

                                                text  label  \
0  NEW YORK (Reuters) - Veteran actor and frequen...      1   
1  BARCELONA (Reuters) - Catalonia s independence...      1   
2  WINSTON-SALEM, N.C. (Reuters) - North Carolina...      1   
3  WASHINGTON (Reuters) - Draft legislation respo...      1   
4  SACRAMENTO, Calif. (Reuters) - California Gove...      1   

                                             content  Predicted Label  
0  Live from New York, it's a Trump-Clinton remat...                1  
1  Catalan separatists to lose majority in tight ...                1  
2  North Carolina governor concedes election to D...                1  
3  Draft Senate Iran legislation sets tough new U...                1  
4  California governor taps U.S. Representative B...                1

In [ ]:
!pip freeze > requirements.txt # menyimpan semua library Python yang sedang terinstall 